src/gui/maptools/qgsbezierdata.cpp

```diff
@@ -0,0 +1,335 @@
+  // Total knots: 4 + 3*(n-2) + 4 = 3n + 2
+  // Actually for n-1 segments with degree 3: ctrlPts.count() + 4 = 3n-2+4 = 3n+2
+  QVector<double> knots;
+
+  // First 4 knots are 0
+  for ( int i = 0; i < 4; ++i )
+    knots.append( 0.0 );
+
+  // Interior knots with multiplicity 3
+  for ( int i = 1; i < n - 1; ++i )
+  {
+    for ( int j = 0; j < 3; ++j )
+      knots.append( static_cast<double>( i ) );
+  }
+
+  // Last 4 knots are n-1
+  for ( int i = 0; i < 4; ++i )
+    knots.append( static_cast<double>( n - 1 ) );
+
+  // Uniform weights (non-rational B-spline)
+  QVector<double> weights( ctrlPts.count(), 1.0 );
+
+  return new QgsNurbsCurve( ctrlPts, 3, knots, weights );
+}
+
+void QgsBezierData::clear()
+{
+  mAnchors.clear();
+  mHandles.clear();
+}
+
+int QgsBezierData::findClosestAnchor( const QgsPoint &pt, double tolerance ) const
+{
+  int closestIdx = -1;
+  double minDistSq = tolerance * tolerance;
+
+  for ( int i = 0; i < mAnchors.count(); ++i )
+  {
+    const double dx = mAnchors[i].x() - pt.x();
+    const double dy = mAnchors[i].y() - pt.y();
+    const double distSq = dx * dx + dy * dy;
+    if ( distSq < minDistSq )
+    {
+      minDistSq = distSq;
+      closestIdx = i;
+    }
+  }
+
+  return closestIdx;
+}
+
+int QgsBezierData::findClosestHandle( const QgsPoint &pt, double tolerance ) const
+{
+  int closestIdx = -1;
+  double minDistSq = tolerance * tolerance;
+
+  for ( int i = 0; i < mHandles.count(); ++i )
+  {
+    // Skip handles that are at anchor position (retracted)
+    const int anchorIdx = i / 2;
+    if ( anchorIdx < mAnchors.count() )
+    {
+      const QgsPoint &anchor = mAnchors[anchorIdx];
+      if ( qFuzzyCompare( mHandles[i].x(), anchor.x() ) && qFuzzyCompare( mHandles[i].y(), anchor.y() ) )
+        continue;
+    }
+
+    const double dx = mHandles[i].x() - pt.x();
+    const double dy = mHandles[i].y() - pt.y();
+    const double distSq = dx * dx + dy * dy;
+    if ( distSq < minDistSq )
+    {
+      minDistSq = distSq;
+      closestIdx = i;
+    }
+  }
+
+  return closestIdx;
+}
+
+int QgsBezierData::findClosestSegment( const QgsPoint &pt, double tolerance ) const
+{
+  if ( mAnchors.count() < 2 )
+    return -1;
+
+  int closestSegment = -1;
+  double minDist = tolerance;
+
+  // Check each segment
+  for ( int i = 0; i < mAnchors.count() - 1; ++i )
+  {
+    const QgsPoint &p0 = mAnchors[i];
+    const QgsPoint &p1 = mHandles[i * 2 + 1];
+    const QgsPoint &p2 = mHandles[( i + 1 ) * 2];
+    const QgsPoint &p3 = mAnchors[i + 1];
+
+    // Sample the curve and find minimum distance
+    for ( int j = 0; j <= INTERPOLATION_POINTS; ++j )
+    {
+      const double t = static_cast<double>( j ) / INTERPOLATION_POINTS;
+      const QgsPoint curvePoint = evaluateBezier( p0, p1, p2, p3, t );
+
+      const double dx = curvePoint.x() - pt.x();
+      const double dy = curvePoint.y() - pt.y();
+      const double dist = std::sqrt( dx * dx + dy * dy );
```

Comment:
very minor optimization possibility that can save the sqrt() if you just compare the square distance to the squared tolerance